crates/hir/src/core/semantic/reference/mod.rs

```diff
@@ -0,0 +1,525 @@
+            return TargetResolution::Single(local);
+        }
+
+        // Fall back to module-level resolution (may be ambiguous)
+        let scopes = resolve_path_to_scopes(db, self.path, self.scope);
+        match scopes.len() {
+            0 => TargetResolution::None,
+            1 => TargetResolution::Single(Target::Scope(scopes.into_iter().next().unwrap())),
+            _ => TargetResolution::Ambiguous(scopes.into_iter().map(Target::Scope).collect()),
+        }
+    }
+
+    /// Resolve this path at a specific cursor position (segment-aware).
+    ///
+    /// Clicking on `foo` in `foo::Bar` resolves to `foo`, not `Bar`.
+    pub fn target_at<DB>(&self, db: &'db DB, cursor: TextSize) -> TargetResolution<'db>
+    where
+        DB: HirAnalysisDb + SpannedHirDb,
+    {
+        let last_idx = self.path.segment_index(db);
+
+        // Find which segment the cursor is in
+        for idx in 0..=last_idx {
+            let Some(seg_span) = self.span.clone().segment(idx).resolve(db) else {
+                continue;
+            };
+
+            if seg_span.range.contains(cursor) {
+                // If this is the last segment, try local binding resolution first
+                // (locals shadow module-level items and are never ambiguous)
+                if idx == last_idx
+                    && let Some(local) = self.local_target(db)
+                {
+                    return TargetResolution::Single(local);
+                }
+
+                // Try module-level resolution for this segment (may be ambiguous)
+                if let Some(seg_path) = self.path.segment(db, idx) {
+                    let scopes = resolve_path_to_scopes(db, seg_path, self.scope);
+                    return match scopes.len() {
+                        0 => TargetResolution::None,
+                        1 => TargetResolution::Single(Target::Scope(
+                            scopes.into_iter().next().unwrap(),
+                        )),
+                        _ => TargetResolution::Ambiguous(
+                            scopes.into_iter().map(Target::Scope).collect(),
+                        ),
+                    };
+                }
+                return TargetResolution::None;
+            }
+        }
+
+        // Cursor not in any segment, try full path resolution
+        self.target(db)
+    }
+
+    /// Resolve to a local binding target (internal helper).
+    fn local_target(&self, db: &'db dyn HirAnalysisDb) -> Option<Target<'db>> {
+        let body = self.scope.body()?;
+        let func = body.containing_func(db)?;
+        let expr_id = body.find_expr_for_path(db, self.path)?;
+        let (_, typed_body) = check_func_body(db, func);
+
+        let span = typed_body.expr_binding_def_span(func, expr_id)?;
+        let ty = typed_body.expr_ty(db, expr_id);
+
+        Some(Target::Local {
+            span,
+            ty,
+            func,
+            expr: expr_id,
+        })
+    }
```

Comment:
PathView as it's defined, and this function (and find_expr_for_path) don't suffice for variable names in function bodies, where scoping rules rule.

Reply:
I'm working on this, still not satisfied though

Reply:
should be addressed now, not in the most elegant way, but effectively